File: app/src/main/java/com/rosberry/android/jigit/presentation/repositories/RepositoriesPresenter.kt
/*
 *
 *  * Copyright (c) 2018 Rosberry. All rights reserved.
 *
 */

package com.rosberry.android.jigit.presentation.repositories

import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import com.rosberry.mpp.jigitbl.domain.RepositoriesInteractor
import com.rosberry.mpp.jigitbl.entity.Repository
import javax.inject.Inject

/**
 * @author Alexei Korshun on 22/02/2019.
 */
@InjectViewState
class RepositoriesPresenter @Inject constructor(
        private val repositoriesInteractor: RepositoriesInteractor
) : MvpPresenter<RepositoriesView>() {

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        viewState.showProgress(true)
        repositoriesInteractor.getRepositories(this::onReceiveRepositories, this::onError)
    }

    private fun onReceiveRepositories(repositories: List<Repository>) {
        viewState.showError(false)
        viewState.showProgress(false)
        viewState.showRepositories(repositories)
    }

    private fun onError(exception: Throwable) {
        exception.printStackTrace()
        viewState.showProgress(false)
        viewState.showError(true, exception.localizedMessage)
    }
}

File: iosApp/jiGit/jiGit/src/presentation/repositroies/RepositoriesPresenter.swift
//
// Created by Aleksei Korshun on 2019-02-22.
// Copyright (c) 2019 Aleksei Korshun. All rights reserved.
//

import Foundation
import FRAMEWORK

class RepositoriesPresenter {

    private let repositoriesInteractor: RepositoriesInteractor

    weak var view: RepositoriesViewController? = nil

    init(_ repositoriesInteractor: RepositoriesInteractor) {
        self.repositoriesInteractor = repositoriesInteractor
    }

    func onAttach() {
        loadRepositories()
    }

    private func loadRepositories() {
        view?.isLoading(isLoading: true)
        repositoriesInteractor.getRepositories(onResult: onRepositoriesLoaded, onError: onRepositoriesFailed)
    }

    private func onRepositoriesLoaded(repositories: [Repository]) {
        view?.isLoading(isLoading: false)
        view?.showRepositories(data: repositories)
    }

    private func onRepositoriesFailed(throwable: KotlinThrowable) {
        view?.isLoading(isLoading: false)
        view?.isError(isError: true, message: throwable.message)
    }
}
